services/queue/README.md

```diff
@@ -21,3 +21,91 @@ For S3 we have a dummy bucket called `test-bucket-for-any-garbage` which stores
 Mozilla developers can get access from a taskcluster developer, or you can setup a custom a bucket and overwrite the bucket name as well as the credentials.
 
 Same thing applies for azure, though it's not as nicely scoped, and doesn't clean up on its own.
+
+## Queue internals
+
+Queue service is using internal queues to manage task lifecycle. Those are:
+
+- pending queue - scheduled tasks ready to be claimed
+- claim queue - tasks that were claimed, and are awaiting to be resolved or failed with `claim-expired`
+- deadline queue - task deadlines, to help resolve task graph dependencies if task was not resolved in time
+- resolve queue - short-lived queue to resolve task graph dependencies
+
+Interactions with those queues are handled in [`QueueService`](./src/queueservice.js).
+
+### Create task
+
+When task is created it can either be scheduled right away (no unmet dependencies) or it can stay unscheduled.
+
+```mermaid
+sequenceDiagram
+  participant client
+  participant queue as Queue API
+  participant deadlineQueue as Deadline Queue
+  participant pendingQueue as Pending Queue
+
+  client->>+queue: createTask()
+  queue-->>deadlineQueue: QueueService.putDeadlineMessage()
+
+  alt is scheduled:
+    queue-->>pendingQueue: if task can be scheduled<br> QueueService.putPendingMessage()
+  end
+
+  queue->>-client: task status response
+```
+
+### Worker claiming work
+
+Workers would be calling `queue.claimWork()` to get tasks from specific `taskQueueId`.
```

Comment:
It might be useful to include reclaimTask in this diagram, as it "bumps" the visibility in the claim queue.

Reply:
Oh! I missed to also check this flow :) Thanks again!

So the `reclaimTask` calls adds new record in claim queue. 
Which means that long running tasks could have multiple records in the claim queue table. 

We can make `(taskId, runId)` unique, and on conflict just update the `takenUntil`. Or delete existing record before.
There is no point in letting ClaimResolver process "dead" messages

Reply:
ah, checked prod db, there are multiple records for the same task, so adding unique key at that time wouldn't help.

Reply:
oh, interesting, some workers reclaim tasks more often, which results in some tasks having `4` claim records for the same `(taskId, runId)` 

Probably worth optimizing it by keeping only the latest one

Reply:
ok, updated migration and function with unique taskId+runId